presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -114,9 +114,19 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
         JdbcTableHandle handle = (JdbcTableHandle) table;
 
         if (handle.getGroupingSets().isPresent()) {
-            // handle's aggregations are applied after constraint, so we cannot apply filter if aggregates is already set
-            // TODO (https://github.com/prestosql/presto/issues/4112) allow filter pushdown after aggregation pushdown
-            return Optional.empty();
+            if (constraint.getSummary().getDomains().isEmpty()) {
+                return Optional.empty();
+            }
+
+            Map<ColumnHandle, Domain> domains = constraint.getSummary().getDomains().get();
+            List<List<JdbcColumnHandle>> groupingSets = handle.getGroupingSets().get();
+
+            boolean canPushDown = domains.keySet().stream().allMatch(predicateColumn ->
+                    groupingSets.stream().allMatch(group -> group.contains(predicateColumn)));
```

Comment:
i think we can pushdown if all the constraint columns are fully contained by every grouping set.
it's not testable right now, since the SPI supoports grouping set pushdown, but the engine actually does not yet

i propose

```
Set<ColumnHandle> constraintColumns = constraint.getSummary().getDomains().orElseThrow().keySet();
List<List<JdbcColumnHandle>> groupingSets = handle.getGroupingSets().get();
verify(!groupingSets.isEmpty());
boolean canPushDown = groupingSets.stream()
        .allMatch(groupingSet -> isContainedBy(constraintColumns, groupingSet));
```

```
private static boolean isContainedBy(Set<ColumnHandle> searchedFor, List<JdbcColumnHandle> container)
    {
        if (searchedFor.size() < 3) {
            return container.containsAll(searchedFor);
        }
        return ImmutableSet.copyOf(container).containsAll(searchedFor);
    }
```

(we can also consider the method overkill and just do the defensive version inline)

Reply:
If I'm reading this correctly the conditions are the same. Right? Every constraint column must be in every group set.

What's the motivation for the defensive version, with the copy?

Reply:
> What's the motivation for the defensive version, with the copy?

avoid O(m*n) when long group by (m) and many predicated column (n)

> If I'm reading this correctly the conditions are the same.

correct. i read the initial code incorrectly though